src/builder/commands/docker.rs

```diff
@@ -0,0 +1,384 @@
+    let registry_host = match registry.split_once(':') {
+        Some((host, _port)) => host,
+        None => registry,
+    };
+    insecure_registries.contains(registry_host)
+}
+
+/// See:
+/// - https://github.com/moby/moby/blob/v20.10.11/pkg/archive/whiteouts.go
+/// - https://github.com/moby/moby/blob/v20.10.11/pkg/archive/diff.go#L131
+#[cfg(feature="containers")]
+fn whiteout_entry_handler(entry: &Entry<Box<dyn Read>>, dst_path: &Path) -> Result<bool, String> {
+    let file_name = dst_path.file_name()
+        .and_then(|fname| fname.to_str());
+    let file_name = if let Some(file_name) = file_name {
+        file_name
+    } else {
+        return Ok(false);
+    };
+
+    if entry.header().entry_type() != EntryType::Regular {
+        return Ok(false);
+    }
+
+    if let Some(whiteout) = file_name.strip_prefix(".wh.") {
+        let dir = dst_path.parent().unwrap();
+        if whiteout == ".wh..opq" {
+            // TODO: Track and keep files that were unpacked from the current archive
+            clean_dir(dir, false)?
+        } else {
+            let mut whiteout_path = dir.to_path_buf();
+            whiteout_path.push(whiteout);
+            safe_remove(&whiteout_path)
+                .map_err(|e| format!("Cannot remove {:?} path: {}", &whiteout_path, e))?;
+        }
+        return Ok(true);
+    }
+
+    Ok(false)
+}
+
+#[cfg(feature="containers")]
+async fn download_and_unpack_image(
+    registry: &str, insecure: bool, image: &str, tag: &str, dst_path: &Path
+) -> Result<(), StepError> {
+    let auth_scope = format!("repository:{}:pull", image);
+    let client = build_client(registry, insecure, &[&auth_scope]).await?;
+
+    println!("Downloading docker image: {}/{}:{}", registry, image, tag);
+    let manifest = client.get_manifest(&image, &tag).await?;
+
+    let layers_digests = manifest.layers_digests(None)?;
+
+    let layers_download_semaphore = Arc::new(
+        Semaphore::new(DOCKER_LAYERS_DOWNLOAD_CONCURRENCY)
+    );
+
+    let mut downloaded_layer_txs = vec!();
+    let mut downloaded_layer_rxs = vec!();
+    for _ in &layers_digests {
+        let (tx, rx) = oneshot::channel();
+        downloaded_layer_rxs.push(rx);
+        downloaded_layer_txs.push(tx);
+    }
+
+    let dst_path = dst_path.to_path_buf();
+    let unpack_task = tokio::spawn(async move {
+        for layer_ch in downloaded_layer_rxs {
+            match layer_ch.await {
+                Ok((digest, layer_path)) => {
+                    let dst_path = dst_path.clone();
+                    if let Err(e) = unpack_layer(digest, layer_path, dst_path).await {
+                        return Err(e);
+                    }
+                }
+                Err(_) => {
+                    // Channel is dropped when download task was cancelled
+                },
+            }
+        }
+        Ok(())
+    });
+
+    let mut layers_tasks = vec!();
+    for (digest, tx) in layers_digests.iter().zip(downloaded_layer_txs.into_iter()) {
```

Comment:
Instead of spawning tasks, it's better to use `FuturesUnordered`, like this:
```
    let layers = layers_digests.iter().zip(downloaded_layer_txs.into_iter())
        .map(|(digest, tx)| {
            async move {
               // same code as in task
            }
        })
        .collect::<FuturesUnordered<_>>()
        .try_collect().await?;
```
This way you don't need channels (results returned right into the collection), and cancellation works seamlesly.

Reply:
> This way you don't need channels

I need channels to notify an unpack task that a corresponding layer was downloaded. Thus unpacking downloaded layers occurs while other layers are still downloading.